src/click-outside.directive.ts

```diff
@@ -118,18 +134,24 @@ export class ClickOutsideDirective implements OnInit, OnChanges, OnDestroy {
       this._ngZone.run(() => this.clickOutside.emit(ev));
 
       if (this.attachOutsideOnClick) {
-        this._events.forEach(e => document.body.removeEventListener(e, this._onClickBody));
+        this._onOutsideClick.next();
       }
     }
   }
 
   private _shouldExclude(target): boolean {
-    for (let excludedNode of this._nodesExcluded) {
-      if (excludedNode.contains(target)) {
-        return true;
-      }
-    }
+    return this._nodesExcluded.some(excludedNode => excludedNode.contains(target));
+  }
+
+  private _listenAll(target:  'window' | 'document' | 'body' | any, ...eventNames: string[]): Observable<Event> {
+    const sources = eventNames.map(eventName => {
+      return new Observable<Event>(observer => this._renderer2.listen(target, eventName, ev => observer.next(ev)));
+    });
 
-    return false;
+    return merge(...sources)
+      .pipe(
+        takeUntil(this._beforeInit),
+        takeUntil(this._onDestroy),
```

Comment:
I'm pretty unfamiliar with some rxjs practices. Could you explain what's going on here?

Reply:
this link will answer both your questions:
http://brianflove.com/2016/12/11/anguar-2-unsubscribe-observables/
takeUntil() is a practice to make sure a subscribe gets unsubscribed for sure when some action occurs.
the action can be defined with Subject, and then when you call Subject.next() Subject.complete() (isoden added this in lines 63-64 of this pull request) you make the action happens, which in turn makes the unsubscribe happen for the subscribe that was registered to this Subject (with takeUntil).

Reply:
can't you just add a `takeUntil(merge(this._beforeInit, this._onDestroy))`? instead of double `takeUntil`